Fix FloorDiv should not generate non integer rationals (due to sympy bug)

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164398

FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. Duo to a bug in sympy. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  Fix 
  https://github.com/pytorch/pytorch/issues/164385, 
  https://github.com/pytorch/pytorch/issues/154996
  https://github.com/pytorch/pytorch/issues/153375
  https://github.com/pytorch/pytorch/issues/164063
and internal user issue.
cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

laithsakka committed:
Fix floorDiv should not generate non int rationals

[ghstack-poisoned]

laithsakka committed:
Update on "Fix floorDiv should not generate non int rationals"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

laithsakka committed:
Update on "Fix FloorDiv should not generate non int rationals"


FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

aorenste commented on torch/utils/_sympy/functions.py:
```diff
@@ -308,6 +311,9 @@ def eval(
     ) -> Optional[sympy.Basic]:
         if base == 0 or modulus == 1:
             return sympy.S.Zero
+        # if "/11" in str(base):
```
debug cruft?

laithsakka committed:
Update on "Fix FloorDiv should not generate non integer rationals"


FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  Fix 
  https://github.com/pytorch/pytorch/issues/164385, 
  https://github.com/pytorch/pytorch/issues/154996
  https://github.com/pytorch/pytorch/issues/153375
  https://github.com/pytorch/pytorch/issues/164063
and internal user issue.
cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

aorenste commented on torch/utils/_sympy/functions.py:
Isn't the bug that it's checking `len(terms) != 0` and not `len(tems) != len(sympy.make_args(base))`? i.e. it's dropping terms?

aorenste replied:
Oh - I see. I didn't look lower and missed that we're still including the remainder...

laithsakka committed:
Update on "Fix FloorDiv should not generate non integer rationals"


FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  Fix 
  https://github.com/pytorch/pytorch/issues/164385, 
  https://github.com/pytorch/pytorch/issues/154996
  https://github.com/pytorch/pytorch/issues/153375
  https://github.com/pytorch/pytorch/issues/164063
and internal user issue.
cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

isuruf commented on torch/utils/_sympy/functions.py:
```diff
@@ -267,8 +267,11 @@ def eval(
                 quotient = term / divisor
-
-                if quotient.is_integer:
+                # sympy can generate a quotient with (1/22)*.... such that quotient.is_integer is True
```
Add https://github.com/sympy/sympy/pull/28442 here?

isuruf replied:
Let's check for `sympy.__version__` and apply this workaround only if the version is less than `1.14.0`

laithsakka replied:
is this just a problem for Mul @isuruf

laithsakka committed:
Update on "Fix FloorDiv should not generate non integer rationals"


FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  Fix 
  https://github.com/pytorch/pytorch/issues/164385, 
  https://github.com/pytorch/pytorch/issues/154996
  https://github.com/pytorch/pytorch/issues/153375
  https://github.com/pytorch/pytorch/issues/164063
and internal user issue.
cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168

[ghstack-poisoned]

jansel reviewed (changes requested):
Failing tests?

isuruf commented on torch/utils/_sympy/functions.py:
```diff
@@ -268,7 +268,18 @@ def eval(
+                # sympy can generate a quotient with (1/22)*.... such that quotient.is_integer is True
+                # FloorDiv should not allow that as output. see
+                quotient_is_integer = None
+                if isinstance(quotient, sympy.Mul) and sympy.__version__ < "1.14.0":
```
`TorchVersion` is just an extension of `Version` that allows comparisons with tuples. Comparing strings like `sympy.__version__ < "1.14.0"` is bound to not work as intended.

laithsakka committed:
Update on "Fix FloorDiv should not generate non integer rationals (due to sympy bug)"


FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  Fix 
  https://github.com/pytorch/pytorch/issues/164385, 
  https://github.com/pytorch/pytorch/issues/154996
  https://github.com/pytorch/pytorch/issues/153375
  https://github.com/pytorch/pytorch/issues/164063
and internal user issue.
cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka commented:
> Failing tests?

those are harmless , fixed with     @unittest.skipUnless(HAS_GPU, "Need GPU for this test")

laithsakka committed:
Update on "Fix FloorDiv should not generate non integer rationals (due to sympy bug)"


FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  Fix 
  https://github.com/pytorch/pytorch/issues/164385, 
  https://github.com/pytorch/pytorch/issues/154996
  https://github.com/pytorch/pytorch/issues/153375
  https://github.com/pytorch/pytorch/issues/164063
and internal user issue.
cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Fix FloorDiv should not generate non integer rationals (due to sympy bug)"


FloorDiv eval have this optimization
```
  # Expands (x + y) // b into x // b + y // b.
  # This only works if floor is an identity, i.e. x / b is an integer.
 ```
 
 Before this PR this optimization would generate a result in an expression like this. 
 ```
Mul(Rational(1, 22), Add(Mul(Integer(24), Symbol('s37', integer=True, positive=True)), Integer(672)), FloorDiv(Mul(Symbol('s14', integer=True, positive=True), Symbol('s46', integer=True, positive=True)), Integer(2016)))
 ```
 
 This is because in sympy an expression can have .is_integer =True yet have 1/22 in it!
 This PR ensure we do not generate that by simply opting out if this optimization if we end 
 up with quotient that have such rational.
  
  Fix 
  https://github.com/pytorch/pytorch/issues/164385, 
  https://github.com/pytorch/pytorch/issues/154996
  https://github.com/pytorch/pytorch/issues/153375
  https://github.com/pytorch/pytorch/issues/164063
and internal user issue.
cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge
